experimental/sql/src/ssa_to_impl.py

```diff
@@ -0,0 +1,137 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from xdsl.ir import Operation, MLContext, Region, Block
+from typing import List, Type, Optional
+from xdsl.dialects.builtin import ArrayAttr, StringAttr, ModuleOp, IntegerAttr
+
+from xdsl.pattern_rewriter import RewritePattern, GreedyRewritePatternApplier, PatternRewriteWalker, PatternRewriter, op_type_rewrite_pattern
+
+import dialects.rel_ssa as RelSSA
+import dialects.rel_impl as RelImpl
+
+# This file contains the rewrite infrastructure to translate the relational SSA
+# dialect to the relational implementation dialect. The current design has a
+# parent class `RelSSARewriter` that contains functions used for several
+# `Rewriter`s. All other `Rewriter`s inherit from that class.
+
+
+@dataclass
+class RelSSARewriter(RewritePattern):
+
+  def convert_datatype(self, type_: RelSSA.DataType) -> RelImpl.DataType:
+    if isinstance(type_, RelSSA.String):
+      return RelImpl.String.get(type_.nullable)
+    if isinstance(type_, RelSSA.Int32):
+      return RelImpl.Int32()
+    raise Exception(
+        f"datatype conversion not yet implemented for {type(type_)}")
+
+  def convert_bag(self, bag: RelSSA.Bag) -> RelImpl.Bag:
+    names = [s.elt_name.data for s in bag.schema.data]
+    types = [self.convert_datatype(s.elt_type) for s in bag.schema.data]
+    return RelImpl.Bag.get(types, names)
+
+  def create_tuple_of_bag(self, bag: RelImpl.Bag) -> RelImpl.Tuple:
+    names = [s.elt_name.data for s in bag.schema.data]
+    types = [s.elt_type for s in bag.schema.data]
+    return RelImpl.Tuple.get(types, names)
+
+
+#===------------------------------------------------------------------------===#
+# Expressions
+#===------------------------------------------------------------------------===#
+
+
+@dataclass
+class LiteralRewriter(RelSSARewriter):
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelSSA.Literal, rewriter: PatternRewriter):
+    rewriter.replace_matched_op(
+        [RelImpl.Literal.get(op.value, self.convert_datatype(op.result.typ))])
+
+
+@dataclass
+class ColumnRewriter(RelSSARewriter):
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelSSA.Column, rewriter: PatternRewriter):
+    # TODO: In this current version, there is only one block_arg, so it is clear
+    # what to access here. What if this is not the case?
```

Comment:
I think this will becomes especially interesting when we have joins in our IR, considering that the join region will have multiple tuple arguments, so we will have to figure out to which tuple a given name belongs. I think this can be written by iterating over all tuples and all names, which I will implement once we actually have a usecase for it.